File: lowestAncestorC++/tst/LCADAG-Test.cpp
#include <gtest/gtest.h>
#include "LCADAG.h"


	
TEST(LCADAGcpp, testNewEdge){
	struct Graph *graph = newGraph(5);
	newEdge(graph,1,2);
	newEdge(graph,1,4);
	newEdge(graph,2,3);
	newEdge(graph,3,4);
	EXPECT_EQ("Invalid vertex entered",newEdge(graph,-1,2));
	EXPECT_EQ("Invalid vertex entered",newEdge(graph,1,7));
	EXPECT_EQ(4,graph->totalEdges);
	EXPECT_EQ(5, graph->totalVertices);
	
}
TEST(LCADAGcpp, testgrapharray){
	struct Graph *graph = newGraph(5);
	newEdge(graph,1,2);
	newEdge(graph,1,4);
	newEdge(graph,2,3);
	newEdge(graph,3,4);
	list<int>::iterator i = (graph->graph[1]).begin(); 
	advance(i, 1); // second element in graph[1]
	EXPECT_EQ(4,*i);

	
}

TEST(LCADAGcpp, testIndegree){
	struct Graph *graph = newGraph(5);
	newEdge(graph,1,2);
	newEdge(graph,2,4);
	newEdge(graph,3,3);
	newEdge(graph,1,3);
	
	EXPECT_EQ(2,graph->indegree[3]); //this test fails once every 3 or so runs, i think it has something to do with memory but don't know enough about c++ so say for certain

}

TEST(LCADAGcpp, testValidVertex){
	struct Graph *graph = newGraph(5);
	EXPECT_EQ(false, isValidVertex(graph,-1));
	EXPECT_EQ(true, isValidVertex(graph,1));

}

TEST(LCADAGcpp, testCycle){
	struct Graph *graph = newGraph(5);
	newEdge(graph,0,1);
	newEdge(graph,1,2);
	newEdge(graph,2,0);
	newEdge(graph,2,3);
	newEdge(graph,3,4);
	checkCycle(graph,0);
	EXPECT_EQ(true,graph->cycleExists);

}

TEST(LCADAGcpp , testFindLCA){
	struct Graph *graph = newGraph(10);
	newEdge(graph,0,1);
	newEdge(graph,1,2);
	newEdge(graph,3,4);
	newEdge(graph,4,5);
	newEdge(graph,3,6);
	newEdge(graph,2,5);
	newEdge(graph,5,7);
	EXPECT_EQ(1, findLCA(graph,7,1));
	EXPECT_EQ(4, findLCA(graph,7,4));
	EXPECT_EQ(3, findLCA(graph,5,6));
	EXPECT_EQ(-1, findLCA(graph,2,4));
	EXPECT_EQ(-1, findLCA(graph,7,-1));

	struct Graph *graph2 = newGraph(10);
	newEdge(graph2,0,1);
	newEdge(graph2,1,2);
	newEdge(graph2,3,4);
	newEdge(graph2,4,5);
	newEdge(graph2,3,6);
	newEdge(graph2,2,5);
	newEdge(graph2,5,7);
	newEdge(graph2,7,0);
	EXPECT_EQ(-1, findLCA(graph2,5,6));
}

int main(int argc, char **argv) {
    testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: lowestAncestor/src/LCADAGTest.java
import static org.junit.Assert.*;

import org.junit.Test;

public class LCADAGTest {

	@Test 
	public void testAddEdge(){
		LCADAG graph = new LCADAG(5);
		graph.addEdge(1, 2);
		graph.addEdge(1, 4);
		graph.addEdge(2, 3);
		graph.addEdge(3, 4);
		assertEquals("Invalid vertex entered", graph.addEdge(-1,2)); // can't add negative node
		assertEquals("Invalid vertex entered", graph.addEdge(1,7)); // 7 outside range
		assertEquals(4,graph.totalEdges);
		assertEquals(5,graph.totalVertices);
	}
	@Test 
	public void testisValidVertex(){
		LCADAG graph = new LCADAG(5);
		
		assertEquals(false, graph.isValidVertex(-1)); 
		assertEquals(true, graph.isValidVertex(1)); 
		
	}
	@Test 
	public void testcheckCycle(){
		LCADAG graph = new LCADAG(10);
		
		graph.addEdge(0, 1);
		graph.addEdge(1, 2);
		graph.addEdge(2, 0);
		graph.addEdge(2, 3);
		graph.addEdge(3, 4);
		graph.checkCycle(0);
		assertEquals(true, graph.cycleExists);
	}
	@Test
	public void testIndegree()
	{
		LCADAG graph = new LCADAG(5);
		
		graph.addEdge(1, 2);
		graph.addEdge(2, 4);
		graph.addEdge(3, 3);
		assertEquals(1, graph.indegree[3]);
	}
	
	@Test
	public void testfindLCA()
	{
		LCADAG graph = new LCADAG(10);
		graph.addEdge(0, 1);
		graph.addEdge(1, 2);
		graph.addEdge(3, 4);
		graph.addEdge(4, 5);
		graph.addEdge(3, 6);
		graph.addEdge(2, 5);
		graph.addEdge(5, 7);
		assertEquals(1, graph.findLCA(7,1));
		assertEquals(4, graph.findLCA(7,4));
		assertEquals(3, graph.findLCA(5,6));
		assertEquals(-1, graph.findLCA(2,4));
		assertEquals(-1, graph.findLCA(7,-1));
		
		LCADAG graph2 = new LCADAG(10);
		graph2.addEdge(0, 1);
		graph2.addEdge(1, 2);
		graph2.addEdge(3, 4);
		graph2.addEdge(4, 5);
		graph2.addEdge(3, 6);
		graph2.addEdge(2, 5);
		graph2.addEdge(5, 7);
		graph2.addEdge(7, 0);//making it cycle
		assertEquals(-1, graph2.findLCA(5,6));//returns -1 because cycle exists

	}
	
	
}